sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -774,6 +774,114 @@ case class Encode(value: Expression, charset: Expression)
 }
 
 /**
+ * Replace all substrings of str that match regexp with rep.
+ *
+ * NOTE: this expression is not THREAD-SAFE, we have to create a instance for each thread.
+ */
+case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expression)
+  extends Expression with ImplicitCastInputTypes with CodegenFallback {
```

Comment:
Add `CodegenFallback` as we have the `mutable state` temporally; we will do the performance benchmark for the codegen,  (complexity V.S. the performance gain)